pallets/phala-world/src/incubation.rs

```diff
@@ -548,6 +667,146 @@ pub mod pallet {
 
 			Ok(Pays::No.into())
 		}
+
+		/// Privileged function to set the collection id for the Shell Parts collection.
+		///
+		/// Parameters:
+		/// - `origin` - Expected Overlord admin account to set the Shell Parts Collection ID
+		/// - `collection_id` - Collection ID of the Shell Parts Collection
+		#[pallet::weight(0)]
+		pub fn set_shell_parts_collection_id(
+			origin: OriginFor<T>,
+			collection_id: CollectionId,
+		) -> DispatchResultWithPostInfo {
+			// Ensure Overlord account makes call
+			let sender = ensure_signed(origin)?;
+			pallet_pw_nft_sale::Pallet::<T>::ensure_overlord(&sender)?;
+			// Shell Parts Collection ID is None or else the collection ID was already set
+			ensure!(
+				ShellPartsCollectionId::<T>::get().is_none(),
+				Error::<T>::ShellPartsCollectionIdAlreadySet
+			);
+			<ShellPartsCollectionId<T>>::put(collection_id);
+
+			Self::deposit_event(Event::ShellPartsCollectionIdSet { collection_id });
+
+			Ok(Pays::No.into())
+		}
+
+		/// Privileged function to set the part chosen by an account for a specific Origin of Shell.
+		///
+		/// Parameters:
+		/// - `origin` - Expected Overlord admin account to set the chosen part to the Origin of Shell
+		/// - `collection_id` - Collection ID of Origin of Shell
+		/// - `nft_id` - NFT ID of the Origin of Shell
+		/// - `shell_part` - Key in StorageDoubleMap of the shell part
+		/// - `chosen_part` - Shell part to be stored in Storage
+		#[pallet::weight(0)]
+		pub fn set_origin_of_shell_chosen_parts(
+			origin: OriginFor<T>,
+			collection_id: CollectionId,
+			nft_id: NftId,
+			shell_part: BoundedVec<u8, T::StringLimit>,
+			chosen_part: ShellPartInfoOf<T>,
+		) -> DispatchResultWithPostInfo {
+			// Ensure Overlord account makes call
+			let sender = ensure_signed(origin)?;
+			pallet_pw_nft_sale::Pallet::<T>::ensure_overlord(&sender)?;
+			// Ensure that the collection is an Origin of Shell Collection
+			ensure!(
+				Self::is_origin_of_shell_collection_id(collection_id),
+				Error::<T>::WrongCollectionId
+			);
+			// Ensure the incubation process has started before setting chosen parts
+			ensure!(
+				HasOriginOfShellStartedIncubation::<T>::get((collection_id, nft_id)),
+				Error::<T>::CannotSetOriginOfShellChosenParts
+			);
+
+			let rarity_type_key = pallet_pw_nft_sale::Pallet::<T>::to_boundedvec_key("rarity")?;
+			let rarity_type_value = pallet_rmrk_core::Properties::<T>::get((
+				collection_id,
+				Some(nft_id),
+				&rarity_type_key,
+			))
+			.ok_or(Error::<T>::RarityTypeNotDetected)?;
+			let rarity_type: RarityType = Decode::decode(&mut rarity_type_value.as_slice())
+				.expect("[rarity] should not fail");
+
+			let mut special_parts_count = Self::get_special_parts_count(chosen_part.clone());
```

Comment:
What's the special parts and why we need to track the count?

Btw, if we save all the parts in a single struct as I proposed before, you can just simply count it on-the-fly efficiently.

Reply:
I noticed that the special part will used to verify the equipping. I'm confused because in my understanding all the setting is done by the overlord, and the overlord will verify all the operations before sending to the blockchain. Then why we need to verify it again?

Reply:
This was added bc the user can submit their updated chosen parts which is executed by Overlord. From there, it would check based on the rarity of the Origin of Shell.
- Legendary: Unlimited Special Parts
- Magic: 5 Special Parts
- Prime: 2 Special Parts

This was implemented to satisfy this requirement, but I can remove this if the frontend will make checks to prevent this before Overlord calls this function.

Reply:
Removed this since frontend will check then Overlord will make call.